JupyterHub R Kernel Test does not work

Hi all,

[test_irkernel.py](https://github.com/jupyter/jupyter_kernel_test/blob/00e3f8fc1139757b1d7b52a8d0b3ff85da74b058/test_irkernel.py) does not work. R Kernel works correct. `test_execute_result` does not work. Looks like test waiting on "[1] 6" result and get just "6". Any ideas how I can fix `test_execute_result` in fastest way?

```
[korniichuk@jupyter ~]$ python3 test_ipykernel.py
/usr/lib/python3.6/site-packages/jupyter_client/manager.py:77: DeprecationWarning: KernelManager._kernel_name_changed is deprecated in traitlets 4.1: use @observe and @unobserve instead.
  def _kernel_name_changed(self, name, old, new):
s...s.ssss..s
======================================================================
FAIL: test_execute_result (__main__.IRKernelTests) (code='1+2+3')
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/lib/python3.6/site-packages/jupyter_kernel_test/__init__.py", line 216, in test_execute_result
    self.assertEqual(output_msgs[0]['msg_type'], 'execute_result')
AssertionError: 'display_data' != 'execute_result'
- display_data
+ execute_result
```
**Environment:**
CentOS 7.4
Linux jupyter 3.10.0-693.5.2.el7.x86_64 #1 SMP Fri Oct 20 20:32:50 UTC 2017 x86_64 x86_64 x86_64 GNU/Linux
Python 3.6.3
R 3.4.1
jupyter (1.0.0)
jupyter-client (5.1.0)
jupyter-console (5.2.0)
jupyter-core (4.3.0)
jupyter-kernel-gateway (2.0.1)
jupyter-kernel-test (0.3)
jupyterhub (0.8.0)

--- a/.github/workflows/tests.yml
+++ b/.github/workflows/tests.yml
@@ -56,9 +56,9 @@ jobs:
           path: |
             ${{ steps.pip-cache.outputs.dir }}
             .mypy_cache
-          key: ${{ runner.os }}-python-${{ matrix.python-version }}-pip-${{ hashFiles('**/requirements*.txt') }}-git-${{ github.sha }}
+          key: ${{ runner.os }}-python-${{ matrix.python-version }}-pip-${{ hashFiles('setup.cfg') }}-git-${{ github.sha }}
           restore-keys: |
-            ${{ runner.os }}-python-${{ matrix.python-version }}-pip-${{ hashFiles('**/requirements*.txt') }}
+            ${{ runner.os }}-python-${{ matrix.python-version }}-pip-${{ hashFiles('setup.cfg') }}
             ${{ runner.os }}-python-${{ matrix.python-version }}-pip-
             ${{ runner.os }}-python
             ${{ runner.os }}-

--- a/.github/workflows/tests.yml
+++ b/.github/workflows/tests.yml
@@ -3,6 +3,8 @@ name: Tests
 on:
   push:
   pull_request:
+  schedule:
+    - cron: "0 0 * * *"
 
 defaults:
   run:

--- a/.github/workflows/tests.yml
+++ b/.github/workflows/tests.yml
@@ -39,6 +39,8 @@ jobs:
         uses: conda-incubator/setup-miniconda@v2
         with:
           auto-update-conda: true
+          environment-file: environment.yml
+          activate-environment: jupyter_kernel_test
           python-version: ${{ matrix.python-version }}
 
       - name: Get pip cache dir
@@ -61,12 +63,9 @@ jobs:
 
       - name: Install dependencies
         run: |
-          conda env create -f environment.yml
-          conda activate jupyter_kernel_test
           pip install -e .[test]
 
       - name: Run the tests
         run: |
-          conda activate jupyter_kernel_test
           python test_ipykernel.py
           python test_irkernel.py

--- a/.github/workflows/tests.yml
+++ b/.github/workflows/tests.yml
@@ -61,9 +61,12 @@ jobs:
 
       - name: Install dependencies
         run: |
+          conda env create -f environment.yml
+          conda activate jupyter_kernel_test
           pip install -e .[test]
 
       - name: Run the tests
         run: |
+          conda activate jupyter_kernel_test
           python test_ipykernel.py
           python test_irkernel.py

--- a/.github/workflows/tests.yml
+++ b/.github/workflows/tests.yml
@@ -4,6 +4,10 @@ on:
   push:
   pull_request:
 
+defaults:
+  run:
+    shell: bash -l {0}
+
 jobs:
   test:
     name: Test
@@ -23,15 +27,19 @@ jobs:
       - name: Checkout
         uses: actions/checkout@v2
 
-      - name: Set up Python ${{ matrix.python-version }}
-        uses: actions/setup-python@v2
+      - name: Cache conda
+        uses: actions/cache@v2
         with:
-          python-version: ${{ matrix.python-version }}
+          path: ~/conda_pkgs_dir
+          key:
+            ${{ runner.os }}-conda-${{ env.CACHE_NUMBER }}-${{
+            hashFiles('environment.yml') }}
 
-      - name: Update pip
-        run: |
-          pip install --upgrade pip
-          pip install --upgrade wheel setuptools
+      - name: Setup conda ${{ matrix.PYTHON_VERSION }}
+        uses: conda-incubator/setup-miniconda@v2
+        with:
+          auto-update-conda: true
+          python-version: ${{ matrix.python-version }}
 
       - name: Get pip cache dir
         id: pip-cache
@@ -56,4 +64,6 @@ jobs:
           pip install -e .[test]
 
       - name: Run the tests
-        run: python test_ipykernel.py
+        run: |
+          python test_ipykernel.py
+          python test_irkernel.py

--- a/LICENSE
+++ b/LICENSE
@@ -1,21 +1,19 @@
-BSD 3-Clause License
-
-Copyright (c) 2017, Project Jupyter Contributors
+Copyright (c) 2015-2021 Project Jupyter Contributors
 All rights reserved.
 
 Redistribution and use in source and binary forms, with or without
 modification, are permitted provided that the following conditions are met:
 
-* Redistributions of source code must retain the above copyright notice, this
-  list of conditions and the following disclaimer.
+1. Redistributions of source code must retain the above copyright notice, this
+   list of conditions and the following disclaimer.
 
-* Redistributions in binary form must reproduce the above copyright notice,
-  this list of conditions and the following disclaimer in the documentation
-  and/or other materials provided with the distribution.
+2. Redistributions in binary form must reproduce the above copyright notice,
+   this list of conditions and the following disclaimer in the documentation
+   and/or other materials provided with the distribution.
 
-* Neither the name of the copyright holder nor the names of its
-  contributors may be used to endorse or promote products derived from
-  this software without specific prior written permission.
+3. Neither the name of the copyright holder nor the names of its
+   contributors may be used to endorse or promote products derived from
+   this software without specific prior written permission.
 
 THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE

--- a/environment.yml
+++ b/environment.yml
@@ -2,8 +2,8 @@ name: jupyter_kernel_test
 channels:
   - conda-forge
 dependencies:
-  r-recommended
-  r-irkernel
-  pip
-  setuptools
-  wheel
+  - r-recommended
+  - r-irkernel
+  - pip
+  - setuptools
+  - wheel

--- a/environment.yml
+++ b/environment.yml
@@ -1,0 +1,9 @@
+name: jupyter_kernel_test
+channels:
+  - conda-forge
+dependencies:
+  r-recommended
+  r-irkernel
+  pip
+  setuptools
+  wheel

--- a/jupyter_kernel_test/__init__.py
+++ b/jupyter_kernel_test/__init__.py
@@ -221,7 +221,6 @@ class KernelTests(TestCase):
                         found = True
                     else:
                         continue
-                    self.assertEqual(msg['msg_type'], 'execute_result')
                     self.assertIn('text/plain', msg['content']['data'])
                     self.assertEqual(msg['content']['data']['text/plain'],
                                     sample['result'])
@@ -241,8 +240,15 @@ class KernelTests(TestCase):
                 self.assertEqual(reply['content']['status'], 'ok')
 
                 self.assertGreaterEqual(len(output_msgs), 1)
-                self.assertEqual(output_msgs[0]['msg_type'], 'display_data')
-                self.assertIn(sample['mime'], output_msgs[0]['content']['data'])
+                found = False
+                for msg in output_msgs:
+                    if msg['msg_type'] == 'display_data':
+                        found = True
+                    else:
+                        continue
+                    self.assertIn(sample['mime'], msg['content']['data'])
+                assert found, 'display_data message not found'
+
 
     # this should match one of the values in code_execute_result
     code_history_pattern = ""

--- a/jupyter_kernel_test/__init__.py
+++ b/jupyter_kernel_test/__init__.py
@@ -214,10 +214,18 @@ class KernelTests(TestCase):
                 self.assertEqual(reply['content']['status'], 'ok')
 
                 self.assertGreaterEqual(len(output_msgs), 1)
-                self.assertEqual(output_msgs[0]['msg_type'], 'execute_result')
-                self.assertIn('text/plain', output_msgs[0]['content']['data'])
-                self.assertEqual(output_msgs[0]['content']['data']['text/plain'],
-                                 sample['result'])
+
+                found = False
+                for msg in output_msgs:
+                    if msg['msg_type'] == 'execute_result':
+                        found = True
+                    else:
+                        continue
+                    self.assertEqual(msg['msg_type'], 'execute_result')
+                    self.assertIn('text/plain', msg['content']['data'])
+                    self.assertEqual(msg['content']['data']['text/plain'],
+                                    sample['result'])
+                assert found, 'execute_result message not found'
 
     code_display_data = []
 

--- a/jupyter_kernel_test/__init__.py
+++ b/jupyter_kernel_test/__init__.py
@@ -3,6 +3,7 @@
 # Copyright (c) Jupyter Development Team.
 # Distributed under the terms of the Modified BSD License.
 
+from time import time
 from unittest import TestCase, SkipTest
 from queue import Empty
 
@@ -63,7 +64,7 @@ class KernelTests(TestCase):
                                  store_history=store_history,
                                  stop_on_error=stop_on_error)
 
-        reply = self.kc.get_shell_msg(timeout=timeout)
+        reply = self.get_non_kernel_info_reply(timeout=timeout)
         validate_message(reply, 'execute_reply', msg_id)
 
         busy_msg = run_sync(self.kc.iopub_channel.get_msg)(timeout=1)
@@ -124,6 +125,13 @@ class KernelTests(TestCase):
 
     completion_samples = []
 
+
+    def get_non_kernel_info_reply(self, timeout=None):
+        while True:
+            reply = self.kc.get_shell_msg(timeout=timeout)
+            if reply['header']['msg_type'] != 'kernel_info_reply':
+                return reply
+
     def test_completion(self):
         if not self.completion_samples:
             raise SkipTest
@@ -131,11 +139,11 @@ class KernelTests(TestCase):
         for sample in self.completion_samples:
             with self.subTest(text=sample['text']):
                 msg_id = self.kc.complete(sample['text'])
-                reply = self.kc.get_shell_msg()
+                reply = self.get_non_kernel_info_reply()
                 validate_message(reply, 'complete_reply', msg_id)
                 if 'matches' in sample:
                     self.assertEqual(set(reply['content']['matches']),
-                                     set(sample['matches']))
+                                    set(sample['matches']))
 
     complete_code_samples = []
     incomplete_code_samples = []
@@ -143,7 +151,7 @@ class KernelTests(TestCase):
 
     def check_is_complete(self, sample, status):
         msg_id = self.kc.is_complete(sample)
-        reply = self.kc.get_shell_msg()
+        reply = self.get_non_kernel_info_reply()
         validate_message(reply, 'is_complete_reply', msg_id)
         if reply['content']['status'] != status:
             msg = "For code sample\n  {!r}\nExpected {!r}, got {!r}."
@@ -263,7 +271,7 @@ class KernelTests(TestCase):
         self.flush_channels()
         msg_id = self.kc.history(**histargs)
 
-        reply = self.kc.get_shell_msg(timeout=timeout)
+        reply = self.get_non_kernel_info_reply(timeout=timeout)
         validate_message(reply, 'history_reply', msg_id)
 
         return reply
@@ -338,7 +346,7 @@ class KernelTests(TestCase):
 
         self.flush_channels()
         msg_id = self.kc.inspect(self.code_inspect_sample)
-        reply = self.kc.get_shell_msg(timeout=TIMEOUT)
+        reply = self.get_non_kernel_info_reply(timeout=TIMEOUT)
         validate_message(reply, 'inspect_reply', msg_id)
 
         self.assertEqual(reply['content']['status'], 'ok')

--- a/jupyter_kernel_test/msgspec_v5.py
+++ b/jupyter_kernel_test/msgspec_v5.py
@@ -162,7 +162,7 @@ schema_fragments['execute_reply'] = {"properties": {
         "additionalProperties": True,
     }},
     "user_expressions": {"type": "object"},
-}}
+}, "required": ["status", "execution_count"]}
 
 schema_fragments['inspect_request'] = {"properties": {
     "code": {"type": "string"},

--- a/jupyter_kernel_test/msgspec_v5.py
+++ b/jupyter_kernel_test/msgspec_v5.py
@@ -251,11 +251,12 @@ schema_fragments["kernel_info_reply"] = {"properties": {
     "implementation_version": {"type": "string"},
     "language_info": {"type": "object"},
     "banner": {"type": "string"},
+    "debugger": {"type": "boolean"},
     "help_links": {"type": "array", "items": {"type": "object", "properties": {
         "text": {"type": "string"},
         "url": {"type": "string"}
     }}}
-}}
+}, "required": ["status", "protocol_version", "implementation", "implementation_version", "language_info", "banner"]}
 
 schema_fragments['shutdown_request'] = {"properties": {
     "restart": {"type": "boolean"},

--- a/test_irkernel.py
+++ b/test_irkernel.py
@@ -28,12 +28,9 @@ class IRKernelTests(jkt.KernelTests):
 
     code_generate_error = "raise"
 
-    code_execute_result = [
-        {'code': "1+2+3", 'result': "[1] 6"}
-    ]
-
     code_display_data = [
         {'code': "plot(iris)", 'mime': "image/png"},
+        {'code': "1+2+3", "mime": "text/plain" }
     ]
 
 

--- a/test_irkernel.py
+++ b/test_irkernel.py
@@ -34,5 +34,10 @@ class IRKernelTests(jkt.KernelTests):
     ]
 
 
+
 if __name__ == '__main__':
+    import os
+    # zip is not available on Windows
+    if os.name == 'nt':
+        IRKernelTests.completion_samples = []
     unittest.main()

